### PR TITLE
local_backend: switch from Protobuf to Thrift

### PR DESCRIPTION
**Pull request by martinvonz**
This migrates the native backend from Protobuf to Thrift since Google's Protobuf team does let us import jj into Google's monorepo if it uses a third-party Protobuf library.

Since the native backend is not supported, I didn't write any migration code for it.

# Checklist

If applicable:
- [x] I have updated `CHANGELOG.md`
- [ ] I have updated the documentation (README.md, docs/, demos/)

### CODE REVIEW

File: lib/src/local_backend_model.thrift [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+struct NormalFile {
+  1: required binary id,
+  2: required bool executable,
+}
+
+struct TreeValue {
+  // Exactly one of these should be set
```
Comment (Ralith) [inline lib/src/local_backend_model.thrift:21]: Union?

**Review by Ralith — CHANGES_REQUESTED** [review]

File: lib/src/local_backend_model.thrift (hunk 1 above)
Comment (martinvonz) [reply lib/src/local_backend_model.thrift]: > Union?

Oops, yes. I wrote this before I learned about it and forgot to update it after. Thanks!

File: lib/src/local_backend_model.thrift [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+struct NormalFile {
+  1: required binary id,
+  2: required bool executable,
+}
+
+union TreeValue {
+  1: optional NormalFile normal_file,
+  2: optional binary symlink_id,
+  3: optional binary tree_id,
+  4: optional binary conflict_id,
```
Comment (Ralith) [inline lib/src/local_backend_model.thrift:24]: I think `optional` can be omitted in unions as implied. Not sure what thrift norms are, though.

**Review by Ralith — APPROVED** [review]

File: lib/src/local_backend_model.thrift (hunk 2 above)
Comment (martinvonz) [reply lib/src/local_backend_model.thrift]: I don't know either, but I agree with removing it, and I skipped it in `op_store_model.thrift`.

**Comment by martinvonz** [comment]
> Google's Protobuf team does let us import jj into Google's monorepo if it uses a third-party Protobuf library

That was meant to say "does *not* let us". However, that has since changed (we are now allowed to use the `protobuf` crate, provided we write some conformance tests for it).

**Pull request merged**